From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Narayan Reddy <narayanr@nvidia.com>
Date: Tue, 7 Sep 2021 00:16:03 +0530
Subject: [PATCH 028/243] NVIDIA: SAUCE: net: phy: enable pause in realtek and
 aquantia PHY

Enable flow control support in aquantia and
realtek phy drivers

Bug 200765586

Change-Id: I1e09e641668bcdc08bd2ad5c2520ca06fdb4d587
Signed-off-by: Narayan Reddy <narayanr@nvidia.com>
Signed-off-by: Revanth Kumar Uppala <ruppala@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2909742
Tested-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>

diff --git a/drivers/net/phy/aquantia/aquantia_main.c b/drivers/net/phy/aquantia/aquantia_main.c
index 6b7009e6d4ff..d71758eb67e6 100644
--- a/drivers/net/phy/aquantia/aquantia_main.c
+++ b/drivers/net/phy/aquantia/aquantia_main.c
@@ -35,6 +35,9 @@
 #define PHY_ID_AQR115C	0x31c31c33
 #define PHY_ID_AQR813	0x31c31cb2
 
+#define MDIO_AN_PAUSE                  BIT(10)
+#define MDIO_AN_ASYM_PAUSE             BIT(11)
+
 #define MDIO_PHYXS_VEND_PROV2			0xc441
 #define MDIO_PHYXS_VEND_PROV2_USX_AN		BIT(3)
 
@@ -802,6 +805,16 @@ static int aqr_gen1_config_init(struct phy_device *phydev)
 	u32 led_idx;
 	int ret;
 
+	/* Advertize flow control */
+	linkmode_set_bit(SUPPORTED_Pause, phydev->supported);
+	linkmode_set_bit(SUPPORTED_Asym_Pause, phydev->supported);
+	linkmode_copy(phydev->advertising, phydev->supported);
+
+	/* Configure flow control */
+	ret = phy_read_mmd(phydev, MDIO_MMD_AN, 0x10);
+	ret |= MDIO_AN_PAUSE | MDIO_AN_ASYM_PAUSE;
+	phy_write_mmd(phydev, MDIO_MMD_AN, 0x10, ret);
+
 	/* Check that the PHY interface type is compatible */
 	if (phydev->interface != PHY_INTERFACE_MODE_SGMII &&
 	    phydev->interface != PHY_INTERFACE_MODE_1000BASEKX &&
diff --git a/drivers/net/phy/realtek/realtek_main.c b/drivers/net/phy/realtek/realtek_main.c
index 05df1a33b0e2..125006892b11 100644
--- a/drivers/net/phy/realtek/realtek_main.c
+++ b/drivers/net/phy/realtek/realtek_main.c
@@ -89,14 +89,6 @@
 #define RTL8211F_LEDCR_MASK			GENMASK(4, 0)
 #define RTL8211F_LEDCR_SHIFT			5
 
-/* RTL8211F(D)(I)-VD-CG CLKOUT configuration is specified via magic values
- * to undocumented register pages. The names here do not reflect the datasheet.
- * Unlike other PHY models, CLKOUT configuration does not go through PHYCR2.
- */
-#define RTL8211FVD_CLKOUT_PAGE			0xd05
-#define RTL8211FVD_CLKOUT_REG			0x11
-#define RTL8211FVD_CLKOUT_EN			BIT(8)
-
 /* RTL8211F RGMII configuration */
 #define RTL8211F_RGMII_PAGE			0xd08
 
@@ -174,8 +166,9 @@ MODULE_AUTHOR("Johnson Leung");
 MODULE_LICENSE("GPL");
 
 struct rtl821x_priv {
-	bool enable_aldps;
-	bool disable_clk_out;
+	u16 phycr1;
+	u16 phycr2;
+	bool has_phycr2;
 	struct clk *clk;
 	/* rtl8211f */
 	u16 iner;
@@ -225,6 +218,8 @@ static int rtl821x_probe(struct phy_device *phydev)
 {
 	struct device *dev = &phydev->mdio.dev;
 	struct rtl821x_priv *priv;
+	u32 phy_id = phydev->drv->phy_id;
+	int ret;
 
 	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
 	if (!priv)
@@ -235,10 +230,24 @@ static int rtl821x_probe(struct phy_device *phydev)
 		return dev_err_probe(dev, PTR_ERR(priv->clk),
 				     "failed to get phy clock\n");
 
-	priv->enable_aldps = of_property_read_bool(dev->of_node,
-						   "realtek,aldps-enable");
-	priv->disable_clk_out = of_property_read_bool(dev->of_node,
-						      "realtek,clkout-disable");
+	ret = phy_read_paged(phydev, RTL8211F_PHYCR_PAGE, RTL8211F_PHYCR1);
+	if (ret < 0)
+		return ret;
+
+	priv->phycr1 = ret & (RTL8211F_ALDPS_PLL_OFF | RTL8211F_ALDPS_ENABLE | RTL8211F_ALDPS_XTAL_OFF);
+	if (of_property_read_bool(dev->of_node, "realtek,aldps-enable"))
+		priv->phycr1 |= RTL8211F_ALDPS_PLL_OFF | RTL8211F_ALDPS_ENABLE | RTL8211F_ALDPS_XTAL_OFF;
+
+	priv->has_phycr2 = !(phy_id == RTL_8211FVD_PHYID);
+	if (priv->has_phycr2) {
+		ret = phy_read_paged(phydev, RTL8211F_PHYCR_PAGE, RTL8211F_PHYCR2);
+		if (ret < 0)
+			return ret;
+
+		priv->phycr2 = ret & RTL8211F_CLKOUT_EN;
+		if (of_property_read_bool(dev->of_node, "realtek,clkout-disable"))
+			priv->phycr2 &= ~RTL8211F_CLKOUT_EN;
+	}
 
 	phydev->priv = priv;
 
@@ -551,11 +560,22 @@ static int rtl8211c_config_init(struct phy_device *phydev)
 			    CTL1000_ENABLE_MASTER | CTL1000_AS_MASTER);
 }
 
-static int rtl8211f_config_rgmii_delay(struct phy_device *phydev)
+static int rtl8211f_config_init(struct phy_device *phydev)
 {
+	struct rtl821x_priv *priv = phydev->priv;
+	struct device *dev = &phydev->mdio.dev;
 	u16 val_txdly, val_rxdly;
 	int ret;
 
+	ret = phy_modify_paged_changed(phydev, RTL8211F_PHYCR_PAGE, RTL8211F_PHYCR1,
+				       RTL8211F_ALDPS_PLL_OFF | RTL8211F_ALDPS_ENABLE | RTL8211F_ALDPS_XTAL_OFF,
+				       priv->phycr1);
+	if (ret < 0) {
+		dev_err(dev, "aldps mode  configuration failed: %pe\n",
+			ERR_PTR(ret));
+		return ret;
+	}
+
 	switch (phydev->interface) {
 	case PHY_INTERFACE_MODE_RGMII:
 		val_txdly = 0;
@@ -585,114 +605,58 @@ static int rtl8211f_config_rgmii_delay(struct phy_device *phydev)
 				       RTL8211F_TXCR, RTL8211F_TX_DELAY,
 				       val_txdly);
 	if (ret < 0) {
-		phydev_err(phydev, "Failed to update the TX delay register: %pe\n",
-			   ERR_PTR(ret));
+		dev_err(dev, "Failed to update the TX delay register\n");
 		return ret;
 	} else if (ret) {
-		phydev_dbg(phydev,
-			   "%s 2ns TX delay (and changing the value from pin-strapping RXD1 or the bootloader)\n",
-			   str_enable_disable(val_txdly));
+		dev_dbg(dev,
+			"%s 2ns TX delay (and changing the value from pin-strapping RXD1 or the bootloader)\n",
+			str_enable_disable(val_txdly));
 	} else {
-		phydev_dbg(phydev,
-			   "2ns TX delay was already %s (by pin-strapping RXD1 or bootloader configuration)\n",
-			   str_enabled_disabled(val_txdly));
+		dev_dbg(dev,
+			"2ns TX delay was already %s (by pin-strapping RXD1 or bootloader configuration)\n",
+			str_enabled_disabled(val_txdly));
 	}
 
 	ret = phy_modify_paged_changed(phydev, RTL8211F_RGMII_PAGE,
 				       RTL8211F_RXCR, RTL8211F_RX_DELAY,
 				       val_rxdly);
 	if (ret < 0) {
-		phydev_err(phydev, "Failed to update the RX delay register: %pe\n",
-			   ERR_PTR(ret));
+		dev_err(dev, "Failed to update the RX delay register\n");
 		return ret;
 	} else if (ret) {
-		phydev_dbg(phydev,
-			   "%s 2ns RX delay (and changing the value from pin-strapping RXD0 or the bootloader)\n",
-			   str_enable_disable(val_rxdly));
+		dev_dbg(dev,
+			"%s 2ns RX delay (and changing the value from pin-strapping RXD0 or the bootloader)\n",
+			str_enable_disable(val_rxdly));
 	} else {
-		phydev_dbg(phydev,
-			   "2ns RX delay was already %s (by pin-strapping RXD0 or bootloader configuration)\n",
-			   str_enabled_disabled(val_rxdly));
+		dev_dbg(dev,
+			"2ns RX delay was already %s (by pin-strapping RXD0 or bootloader configuration)\n",
+			str_enabled_disabled(val_rxdly));
 	}
 
-	return 0;
-}
-
-static int rtl8211f_config_clk_out(struct phy_device *phydev)
-{
-	struct rtl821x_priv *priv = phydev->priv;
-	int ret;
-
-	/* The value is preserved if the device tree property is absent */
-	if (!priv->disable_clk_out)
+	if (!priv->has_phycr2)
 		return 0;
 
-	if (phydev->drv->phy_id == RTL_8211FVD_PHYID)
-		ret = phy_modify_paged(phydev, RTL8211FVD_CLKOUT_PAGE,
-				       RTL8211FVD_CLKOUT_REG,
-				       RTL8211FVD_CLKOUT_EN, 0);
-	else
-		ret = phy_modify_paged(phydev, RTL8211F_PHYCR_PAGE,
-				       RTL8211F_PHYCR2, RTL8211F_CLKOUT_EN, 0);
-	if (ret)
-		return ret;
-
-	return genphy_soft_reset(phydev);
-}
-
-/* Advance Link Down Power Saving (ALDPS) mode changes crystal/clock behaviour,
- * which causes the RXC clock signal to stop for tens to hundreds of
- * milliseconds.
- *
- * Some MACs need the RXC clock to support their internal RX logic, so ALDPS is
- * only enabled based on an opt-in device tree property.
- */
-static int rtl8211f_config_aldps(struct phy_device *phydev)
-{
-	struct rtl821x_priv *priv = phydev->priv;
-	u16 mask = RTL8211F_ALDPS_PLL_OFF |
-		   RTL8211F_ALDPS_ENABLE |
-		   RTL8211F_ALDPS_XTAL_OFF;
-
-	/* The value is preserved if the device tree property is absent */
-	if (!priv->enable_aldps)
-		return 0;
-
-	return phy_modify_paged(phydev, RTL8211F_PHYCR_PAGE, RTL8211F_PHYCR1,
-				mask, mask);
-}
-
-static int rtl8211f_config_phy_eee(struct phy_device *phydev)
-{
 	/* Disable PHY-mode EEE so LPI is passed to the MAC */
-	return phy_modify_paged(phydev, RTL8211F_PHYCR_PAGE, RTL8211F_PHYCR2,
-				RTL8211F_PHYCR2_PHY_EEE_ENABLE, 0);
-}
-
-static int rtl8211f_config_init(struct phy_device *phydev)
-{
-	struct device *dev = &phydev->mdio.dev;
-	int ret;
-
-	ret = rtl8211f_config_aldps(phydev);
-	if (ret) {
-		dev_err(dev, "aldps mode configuration failed: %pe\n",
-			ERR_PTR(ret));
-		return ret;
-	}
-
-	ret = rtl8211f_config_rgmii_delay(phydev);
+	ret = phy_modify_paged(phydev, RTL8211F_PHYCR_PAGE, RTL8211F_PHYCR2,
+			       RTL8211F_PHYCR2_PHY_EEE_ENABLE, 0);
 	if (ret)
 		return ret;
 
-	ret = rtl8211f_config_clk_out(phydev);
-	if (ret) {
+	ret = phy_modify_paged(phydev, RTL8211F_PHYCR_PAGE,
+			       RTL8211F_PHYCR2, RTL8211F_CLKOUT_EN,
+			       priv->phycr2);
+	if (ret < 0) {
 		dev_err(dev, "clkout configuration failed: %pe\n",
 			ERR_PTR(ret));
 		return ret;
 	}
 
-	return rtl8211f_config_phy_eee(phydev);
+	/* Advertise Flow Control */
+	linkmode_set_bit(SUPPORTED_Pause, phydev->supported);
+	linkmode_set_bit(SUPPORTED_Asym_Pause, phydev->supported);
+	linkmode_copy(phydev->advertising, phydev->supported);
+
+	return genphy_soft_reset(phydev);
 }
 
 static int rtl821x_suspend(struct phy_device *phydev)
-- 
2.43.0

